rust/driver/src/lib.rs

```diff
@@ -42,255 +46,319 @@ pub unsafe extern "C" fn get_tx_context_zeroed(_context: *mut ffi::c_void) -> ev
-    revision: Revision,
-    message: &evmc_message,
-    code: &[u8],
-    status: evmc_step_status_code,
-    pc: u64,
-    gas_refunds: i64,
-    stack: &mut [Uint256],
-    memory: &mut [u8],
-    last_call_result_data: &mut [u8],
-    steps: i32,
-) -> evmc_step_result {
-    // SAFETY:
-    // All pointer are valid since they are created from references.
-    unsafe {
-        run_steppable_raw(
-            host,
-            context as *mut T as *mut ffi::c_void,
-            revision,
-            message,
-            if code.is_empty() {
-                ptr::null()
-            } else {
-                code.as_ptr()
-            },
-            code.len(),
-            status,
-            pc,
-            gas_refunds,
-            if stack.is_empty() {
-                ptr::null_mut()
-            } else {
-                stack.as_mut_ptr()
-            },
-            stack.len(),
-            if memory.is_empty() {
-                ptr::null_mut()
-            } else {
-                memory.as_mut_ptr()
-            },
-            memory.len(),
-            if last_call_result_data.is_empty() {
-                ptr::null_mut()
-            } else {
-                last_call_result_data.as_mut_ptr()
-            },
-            last_call_result_data.len(),
-            steps,
-        )
+impl Drop for SteppableInstance {
+    fn drop(&mut self) {
+        let destroy = self.0.destroy.unwrap();
+        unsafe { destroy(self.0) };
     }
 }
 
-#[allow(clippy::too_many_arguments)]
-pub fn run_steppable_with_null_context(
-    host: &evmc_host_interface,
-    revision: Revision,
-    message: &evmc_message,
-    code: &[u8],
-    status: evmc_step_status_code,
-    pc: u64,
-    gas_refunds: i64,
-    stack: &mut [Uint256],
-    memory: &mut [u8],
-    last_call_result_data: &mut [u8],
-    steps: i32,
-) -> evmc_step_result {
-    // SAFETY:
-    // All pointer are valid since they are created from references except for `context` which is
-    // allowed to be null.
-    unsafe {
-        run_steppable_raw(
-            host,
-            ptr::null_mut(),
-            revision,
-            message,
-            if code.is_empty() {
-                ptr::null()
-            } else {
-                code.as_ptr()
-            },
-            code.len(),
-            status,
-            pc,
-            gas_refunds,
-            if stack.is_empty() {
-                ptr::null_mut()
-            } else {
-                stack.as_mut_ptr()
-            },
-            stack.len(),
-            if memory.is_empty() {
-                ptr::null_mut()
-            } else {
-                memory.as_mut_ptr()
-            },
-            memory.len(),
-            if last_call_result_data.is_empty() {
-                ptr::null_mut()
-            } else {
-                last_call_result_data.as_mut_ptr()
-            },
-            last_call_result_data.len(),
-            steps,
-        )
+impl SteppableInstance {
```

Comment:
The three implementations here seem redundant/too complicated to me, I would try to make an effort to have a single entry point into the VM. The minimum we can do now is document precisely why do they exist.